### PR TITLE
Generic encoding

### PR DESCRIPTION
**Pull request by blinky3713**
This is a really big change but it applies what we learned from purescript-web3 to fix a lot of bugs, namely #29 #28 #26 #25 and a lot of #8. 

1. change the template haskell encoding to use generics, specifically generics-sop. This fixes the de/encoding problem with tuples. The reason for doing this is that it's much easier to reason about to build up the encodings generically like this, and it's also needed for the way that the events decoding works. 

2. create proper types for `uint`, `int`, and sized vector types and fix all the encoding problems there, use these types in the quasi quoter.

3. correctly implement event decoding. This also involved splitting out the `ABIEncoding` class to two classes `ABIEncode` and `ABIDecode`, similar to other encoding libraries. Event encoding is a pretty subtle problem to solve, because it relies on being able to merge to tuples together in the right order. This is why the type level merge sort and `Tagged` types were used. Also we can now use anonymous events.

4. write tests for all of this, both unit tests and e2e tests that work with the existing e2e testing framework that @iostat wrote. 

5. change the generated event types to use records rather than simple products. this is particularly useful with the `RecordWildcards` extension as you don't have to remember the order of arguments. I had to make some simple assumptions for generating the record field names. 

I'm expecting that there will be some changes here, but let's get the conversation started i guess

### CODE REVIEW

**Comment by akru** [comment]
Looks great, thank you @blinky3713!
I review it as soon as possible.

**Comment by blinky3713** [comment]
Ok let’s discuss it after you have read, I have a few comments

**Comment by akru** [comment]
@blinky3713 can you see this: https://travis-ci.org/airalab/hs-web3/jobs/324695503#L717

**Comment by blinky3713** [comment]
It looks like that build is using ghc 7.10, but you’re not specifying a compiler version anywhere in this project. I believe I was using 8.2.2 to compile this, can we try upgrading the Travis job ?

**Comment by blinky3713** [comment]
Oh yeah, weird I just tried a clean build and it failed for the same reason. I don't know why that failed before.

File: src/Network/Ethereum/Web3/Contract.hs [hunk 1]
```diff
@@ -62,29 +77,20 @@ data EventAction = ContinueEvent
...
-          -- ^ Contract address
-          -> (a -> ReaderT Change (Web3 p) EventAction)
-          -- ^ 'Event' handler
-          -> Web3 p ThreadId
-          -- ^ 'Web3' wrapped event handler spawn ident
-    event = _event
+    eventFilter :: Proxy e -> Address -> Filter
 
-_event :: (Provider p, Event a)
+event :: forall p e i ni.
```
Comment (akru) [inline src/Network/Ethereum/Web3/Contract.hs]: Documentation

File: src/Network/Ethereum/Web3/Contract.hs [hunk 2]
```diff
@@ -96,45 +102,47 @@ _event a f = do
...
-         => Call
-         -- ^ Call configuration
-         -> DefaultBlock
-         -- ^ State mode for constant call (latest or pending)
-         -> a
-         -- ^ Method data
-         -> Web3 p b
-         -- ^ 'Web3' wrapped result
-    call = _call
+class Method a where
```
Comment (akru) [inline src/Network/Ethereum/Web3/Contract.hs]: Documentation

File: src/Network/Ethereum/Web3/Encoding/Event.hs [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE GADTs                  #-}
+{-# LANGUAGE MultiParamTypeClasses  #-}
+{-# LANGUAGE PolyKinds              #-}
+{-# LANGUAGE ScopedTypeVariables    #-}
+{-# LANGUAGE TypeFamilies           #-}
+{-# LANGUAGE TypeInType             #-}
+{-# LANGUAGE TypeOperators          #-}
+{-# LANGUAGE UndecidableInstances   #-}
+
```
Comment (akru) [inline src/Network/Ethereum/Web3/Encoding/Event.hs]: Documentation header

File: src/Network/Ethereum/Web3/Encoding/Event.hs [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+import           Network.Ethereum.Web3.Address           (Address)
+import           Network.Ethereum.Web3.Encoding          (ABIDecode, fromData)
+import           Network.Ethereum.Web3.Encoding.Generic  (GenericABIDecode,
+                                                          genericFromData)
+import           Network.Ethereum.Web3.Encoding.Internal
+import           Network.Ethereum.Web3.Types             (Change (..))
+
+
+class ArrayParser a where
```
Comment (akru) [inline src/Network/Ethereum/Web3/Encoding/Event.hs]: Documentation (optional)

File: src/Network/Ethereum/Web3/Encoding/Event.hs [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+  decodeEvent change = do
+      let anonymous = isAnonymous (Proxy :: Proxy e)
+      (Event i ni :: Event i ni) <- parseChange change anonymous
+      return $ combineChange i ni
+
+--------------------------------------------------------------------------------
+-- Event Parsing Internals
+--------------------------------------------------------------------------------
+
+data HList :: [*] -> * where
```
Comment (akru) [inline src/Network/Ethereum/Web3/Encoding/Event.hs]: So, may be this type and sorting classes can be moved to separate module from `Event` module?

File: src/Network/Ethereum/Web3/Encoding/Generic.hs [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+{-# LANGUAGE DeriveGeneric       #-}
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE FlexibleInstances   #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving  #-}
+{-# LANGUAGE TypeFamilies        #-}
+{-# LANGUAGE TypeInType          #-}
+{-# LANGUAGE TypeOperators       #-}
+
```
Comment (akru) [inline src/Network/Ethereum/Web3/Encoding/Generic.hs]: Documentation header

File: src/Network/Ethereum/Web3/Encoding/Int.hs [hunk 7]
```diff
@@ -0,0 +1,88 @@
+{-# LANGUAGE DataKinds                  #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE KindSignatures             #-}
+{-# LANGUAGE OverloadedStrings          #-}
+{-# LANGUAGE PolyKinds                  #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+
```
Comment (akru) [inline src/Network/Ethereum/Web3/Encoding/Int.hs]: Documentation header

File: src/Network/Ethereum/Web3/Encoding/Vector.hs [hunk 8]
```diff
@@ -0,0 +1,33 @@
+{-# LANGUAGE DataKinds            #-}
+{-# LANGUAGE FlexibleInstances    #-}
+{-# LANGUAGE KindSignatures       #-}
+{-# LANGUAGE PolyKinds            #-}
+{-# LANGUAGE ScopedTypeVariables  #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
```
Comment (akru) [inline src/Network/Ethereum/Web3/Encoding/Vector.hs]: Documentation header

File: test/Network/Ethereum/Web3/Test/ComplexStorageSpec.hs [hunk 9]
```diff
@@ -0,0 +1,88 @@
+{-# LANGUAGE QuasiQuotes     #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE DataKinds #-}
+
```
Comment (akru) [inline test/Network/Ethereum/Web3/Test/ComplexStorageSpec.hs]: Documentation header with short description about this test case

File: test/Network/Ethereum/Web3/Test/SimpleStorageSpec.hs [hunk 10]
```diff
@@ -1,6 +1,9 @@
 {-# LANGUAGE KindSignatures  #-}
+{-# LANGUAGE LambdaCase  #-}
 {-# LANGUAGE QuasiQuotes     #-}
 {-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+
```
Comment (akru) [inline test/Network/Ethereum/Web3/Test/SimpleStorageSpec.hs]: Documentation header

File: unit/Network/Ethereum/Web3/Test/EventSpec.hs [hunk 11]
```diff
@@ -0,0 +1,78 @@
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE DeriveGeneric #-}
+
```
Comment (akru) [inline unit/Network/Ethereum/Web3/Test/EventSpec.hs]: Documentation header

**Comment by blinky3713** [comment]
@akru I submitted the changes you requested, mostly module documentation. It passes 3 of your 4 tests, and fails the remaining for some configuration reason I don't understand. Also, why are you building on GHC 8.0.1 and 8.0.2 and similarly 8.2.1 and 8.2.2. It seems like sort of a waste if you trust the GHC versioning schema.

File: web3.cabal [hunk 12]
```diff
@@ -48,6 +48,7 @@ library
                      , Network.Ethereum.Web3.Contract
                      , Network.Ethereum.Web3.Encoding.Bytes
                      , Network.Ethereum.Web3.Encoding.Event
+                     , Network.Ethereum.Web3.Encoding.Event.Internal
```
Comment (akru) [inline web3.cabal]: Is is really needed to export?

**Comment by akru** [comment]
About ghc versioning: Travis parallel build cases is cheap and in some Linux distros available ghc-8.0.1 only, for example. With ghc-8.2 similar case. Differences between patch versions is minimal, but if it can be tested cheap - do it.

**Pull request merged**